Intersect.Client.Core/Entities/Resource.cs

```diff
@@ -175,15 +229,48 @@ private void OnAnimationDisposedOrFinished(IAnimation animation)
         animation.Finished -= OnAnimationDisposedOrFinished;
     }
 
-    private void UpdateFromDescriptor(ResourceDescriptor? descriptor)
+    public void UpdateCurrentState()
     {
-        if (descriptor == null)
+        if (Descriptor == default)
         {
             return;
         }
 
-        var updatedSprite = IsDead ? descriptor.Exhausted.Graphic : descriptor.Initial.Graphic;
-        _sprite = updatedSprite;
+        var graphicStates = Descriptor.States;
+        var currentHealthPercentage = Math.Floor((float)Vital[(int)Enums.Vital.Health] / _maximumHealthForStates * 100);
+
+        if (_currentState is { } currentState &&
+            currentHealthPercentage >= _currentState?.MinimumHealth &&
+            currentHealthPercentage <= _currentState?.MaximumHealth
+        )
+        {
+            return;
+        }
+
+        currentState = graphicStates.Values.FirstOrDefault(
+            s => currentHealthPercentage >= s.MinimumHealth && currentHealthPercentage <= s.MaximumHealth
+        );
+
+        // dispose animation if animation is not used in this current state,
+        // but the previous state was an animation
+        if (
+            _currentState?.TextureType == ResourceTextureSource.Animation &&
+            _stateAnimation != default &&
+            currentState?.TextureType != ResourceTextureSource.Animation
+            )
+        {
+            _stateAnimation.Dispose();
+            _stateAnimation = default;
+        }
+
+        _currentState = currentState;
+        _sprite = _currentState?.TextureName ?? string.Empty;
```

Comment:
I would prefer we set `_sprite` to `null` if at all possible, rather than Empty

Reply:
then this would have to be left for another refactoring, since _sprite is defined as not null in entity, and this affects the Sprite property which is also not null, which would require adding several other checks in loadtextures, sprite set, and so on